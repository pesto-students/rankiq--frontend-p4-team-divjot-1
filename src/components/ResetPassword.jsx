import { useEffect, useState } from 'react';
import { useNavigate, useSearchParams } from 'react-router-dom';
import { useController, useForm } from 'react-hook-form';
// @mui
import { useTranslation } from 'react-i18next';
import Container from '@mui/material/Container';
import Grid from '@mui/material/Grid';
import Typography from '@mui/material/Typography';
import { styled } from '@mui/material/styles';
import TextField from '@mui/material/TextField';
import Button from '@mui/material/Button';
import Card from '@mui/material/Card';
import Alert from '@mui/material/Alert';
import Snackbar from '@mui/material/Snackbar';
import CircularProgress from '@mui/material/CircularProgress';
// redux and services
// import { isEmpty } from 'lodash';
import { REGEX } from '../constants';
import { SERVER_BASE_API, USER_API } from '../constants/endpoints';

const StyledCard = styled(Card)(({ theme }) => ({
  backgroundColor: theme.palette.box.main,
}));

function ResetPassword() {
  const { t } = useTranslation();
  const navigate = useNavigate();
  const [searchParams] = useSearchParams();
  const token = searchParams.get('key');

  const [error, setError] = useState(false);
  const [success, setSuccess] = useState(false);
  const [loading, setLoading] = useState(false);
  const [errorMsg, setErrorMsg] = useState('');

  const { control, formState, reset, getValues } = useForm({
    mode: 'onChange',
    defaultValues: { newPassword: '', confirmPassword: '' },
  });

  const validateConfirmPassword = () => {
    const formvalues = getValues();
    if (formvalues.newPassword === formvalues.confirmPassword) {
      return true;
    }
    return false;
  };

  const {
    field: { ref: newPasswordRef, ...newPasswordProps },
    fieldState: { invalid: newPasswordInvalid, error: newPasswordError },
  } = useController({
    name: 'newPassword',
    control,
    rules: {
      required: t('errors.required'),
      pattern: { value: REGEX.PASSWORD, message: t('errors.password') },
    },
  });

  const {
    field: { ref: confirmPasswordRef, ...confirmPasswordProps },
    fieldState: { invalid: confirmPasswordInvalid },
  } = useController({
    name: 'confirmPassword',
    control,
    rules: {
      required: t('errors.required'),
      validate: (value) => validateConfirmPassword(value),
    },
  });

  const handleReset = () => {
    setError(false);
    setError('');
    reset();
  };

  const handleUpdatePassword = () => {
    setLoading(true);
    const data = {
      resetToken: token,
      password: getValues('newPassword'),
    };
    fetch(`${SERVER_BASE_API}${USER_API.UPDATE_PASSWORD}`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(data),
    }).then((res) => {
      if (!res.ok || res.status >= 400) {
        setError(true);
        setErrorMsg('Reset password failed please try again later.');
      } else {
        setSuccess(true);
      }
    });
  };

  const handleClose = (isSuccess) => {
    if (isSuccess) {
      setSuccess(false);
    } else {
      setError(false);
    }
    navigate('/login');
    setLoading(false);
  };

  return (
    <Container maxWidth="sm">
      <StyledCard
        sx={{
          padding: { xs: '2rem 1rem', sm: '2rem' },
        }}
      >
        <form>
          <Grid container spacing={4}>
            <Grid item xs={12}>
              <Typography variant="h6" textAlign="center">
                {t('resetPassword.newPassword')}
              </Typography>
            </Grid>
            <Grid item xs={12}>
              <TextField
                {...newPasswordProps}
                inputRef={newPasswordRef}
                error={newPasswordInvalid}
                label={t('resetPassword.newPassword')}
                type="password"
                helperText={
                  newPasswordError?.message ?? 'please enter your new password.'
                }
                required
                fullWidth
              />
            </Grid>
            <Grid item xs={12}>
              <TextField
                {...confirmPasswordProps}
                inputRef={confirmPasswordRef}
                error={confirmPasswordInvalid}
                label={t('resetPassword.confirmPassword')}
                helperText={
                  confirmPasswordInvalid
                    ? 'passwords do not match'
                    : 'please enter your new password again'
                }
                type="password"
                required
                fullWidth
              />
            </Grid>
            <Grid item xs={6}>
              <Button
                size="large"
                variant="contained"
                disabled={!formState.isValid || loading}
                fullWidth
                onClick={handleUpdatePassword}
              >
                {loading ? (
                  <CircularProgress size={25} />
                ) : (
                  t('resetPassword.updatePassword')
                )}
              </Button>
            </Grid>

            <Grid item xs={6}>
              <Button
                size="large"
                variant="outlined"
                onClick={handleReset}
                fullWidth
              >
                {t('generic.reset')}
              </Button>
            </Grid>
          </Grid>
        </form>
      </StyledCard>

      <Snackbar
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center',
        }}
        open={error}
        autoHideDuration={6000}
        onClose={() => handleClose(false)}
      >
        <Alert severity="error">{errorMsg}</Alert>
      </Snackbar>
      <Snackbar
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center',
        }}
        open={success}
        autoHideDuration={6000}
        onClose={() => handleClose(true)}
      >
        <Alert severity="success">Password updated successfully</Alert>
      </Snackbar>
    </Container>
  );
}

export default ResetPassword;
