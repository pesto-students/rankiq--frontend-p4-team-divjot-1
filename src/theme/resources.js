import enDefault from './l10n/en.json';
import hiDefault from './l10n/hi.json';

const resources = {
  en: {
    default: enDefault,
  },
  hi: {
    default: hiDefault,
  },
};

export default resources;
