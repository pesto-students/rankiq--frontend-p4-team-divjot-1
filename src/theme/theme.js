import { createTheme } from '@mui/material/styles';

const themeOptions = {
  palette: {
    type: 'light',
    primary: {
      main: '#3f51b5',
    },
    secondary: {
      main: '#ffffff',
    },
    text: { navFooter: '#FFFFFF' },
    box: { main: '#FAFAFA' },
  },
};

const theme = createTheme(themeOptions);

export default theme;
