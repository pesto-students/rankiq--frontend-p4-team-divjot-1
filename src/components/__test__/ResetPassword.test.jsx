import { beforeEach, describe, expect, it, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';
import { ThemeProvider } from '@mui/material/styles';
import { BrowserRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import ResetPassword from '../ResetPassword';
import theme from '../../theme/theme';

const mockStore = configureStore([thunk]);
const navigate = vi.fn();
vi.mock('react-router-dom', async () => {
  return {
    ...(await vi.importActual('react-router-dom')),
    useNavigate: () => navigate,
  };
});

describe('Dashboard.jsx', () => {
  let tree;
  let store;
  const storeData = {
    auth: {
      userInfo: { email: 'email', firstName: 'Test Name' },
      accessToken: 'testToken',
    },
    userHistory: {
      loading: false,
      data: [
        { rollNumber: '1234', mark: 50 },
        { rollNumber: '4567', mark: 11 },
      ],
    },
    rankData: {
      loading: false,
      error: null,
      data: {
        overall: { rank: 1, overall: 12 },
        zone: { rank: 1, overall: 12 },
        zoneCategory: { rank: 1, overall: 12 },
        zoneShift: { rank: 1, overall: 12 },
        zoneShiftCategory: { rank: 1, overall: 12 },
        rawMarks: 12.12,
        zoneSelected: 'test zone',
      },
    },
  };
  beforeEach(() => {
    store = mockStore(storeData);
    tree = render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <BrowserRouter>
            <ResetPassword />
          </BrowserRouter>
        </ThemeProvider>
      </Provider>
    );
  });

  it('should render without any errors', () => {
    const { asFragment } = tree;
    expect(asFragment()).toMatchSnapshot();
  });

  it('should not show any errors when password and confirm password is entered', async () => {
    const user = userEvent.setup();
    const passwordField = screen.getByTestId('ippassword');
    const passwordIp = passwordField.getElementsByTagName('input')[0];
    const cPasswordField = screen.getByTestId('ipcpassword');
    const cPasswordInput = cPasswordField.getElementsByTagName('input')[0];
    await user.click(cPasswordInput);
    await user.keyboard('rankiq@123');
    await user.click(passwordIp);
    await user.keyboard('rankiq@123');
    await user.tab();
    expect(screen.getByTestId('btnSubmit')).not.toBeDisabled();
  });

  it('should not show any errors when password and confirm password is entered', async () => {
    const user = userEvent.setup();
    const passwordField = screen.getByTestId('ippassword');
    const passwordIp = passwordField.getElementsByTagName('input')[0];
    const cPasswordField = screen.getByTestId('ipcpassword');
    const cPasswordInput = cPasswordField.getElementsByTagName('input')[0];
    await user.click(cPasswordInput);
    await user.keyboard('rankiq@123');
    await user.click(passwordIp);
    await user.keyboard('rankiq@123');
    await user.click(screen.getByTestId('btnReset'));
    expect(screen.getByTestId('btnSubmit')).toBeDisabled();
  });
});
