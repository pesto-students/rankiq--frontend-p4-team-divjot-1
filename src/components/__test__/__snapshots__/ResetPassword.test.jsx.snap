// Vitest Snapshot v1

exports[`Dashboard.jsx > should render without any errors 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthSm css-cuefkz-MuiContainer-root"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1or43t3-MuiPaper-root-MuiCard-root"
    >
      <form>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 css-480o17-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-1kgxqm0-MuiTypography-root"
            >
              resetPassword.newPassword
            </h6>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
              data-testid="ippassword"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-1iuujq-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for=":r0:"
                id=":r0:-label"
              >
                resetPassword.newPassword
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-xurl45-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-describedby=":r0:-helper-text"
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r0:"
                  name="newPassword"
                  required=""
                  type="password"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      resetPassword.newPassword *
                    </span>
                  </legend>
                </fieldset>
              </div>
              <p
                class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained Mui-required css-1wc848c-MuiFormHelperText-root"
                id=":r0:-helper-text"
              >
                resetPassword.resetPasswordLinkSent
              </p>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
              data-testid="ipcpassword"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-1iuujq-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for=":r1:"
                id=":r1:-label"
              >
                resetPassword.confirmPassword
                <span
                  aria-hidden="true"
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-xurl45-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-describedby=":r1:-helper-text"
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r1:"
                  name="confirmPassword"
                  required=""
                  type="password"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      resetPassword.confirmPassword *
                    </span>
                  </legend>
                </fieldset>
              </div>
              <p
                class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained Mui-required css-1wc848c-MuiFormHelperText-root"
                id=":r1:-helper-text"
              >
                please enter your new password again
              </p>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth css-fjh6yo-MuiButtonBase-root-MuiButton-root"
              data-testid="btnSubmit"
              disabled=""
              tabindex="-1"
              type="button"
            >
              resetPassword.updatePassword
            </button>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-fullWidth MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-fullWidth css-i8s7vf-MuiButtonBase-root-MuiButton-root"
              data-testid="btnReset"
              tabindex="0"
              type="button"
            >
              generic.reset
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
